### PR TITLE
Add inventory plugin iocage

### PR DESCRIPTION
**Pull request by vbotka**
##### SUMMARY

- Implement  #9261

- Feature Pull Request

- iocage

- Tested in:
   - FreeBSD 14.1-RELEASE
   - FreeBSD 13.1-RELEASE-p9

### CODE REVIEW

File: plugins/inventory/iocage.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+            cmd.append(f"ssh {user}@{host}")
+        if env:
+            cmd.append(' '.join(env))
+        cmd.append(self.IOCAGE)
+
+        cmd_list = cmd.copy()
+        cmd_list.append('list --header --long')
+        cmd_list = ' '.join(cmd_list)
+        try:
+            p = Popen(cmd_list, shell=True, text=True, stdout=PIPE, stderr=PIPE)
```
Comment (russoz) [inline plugins/inventory/iocage.py:186]: We strongly recommend not using shell mode when running Popen, and passing the arguments as a list, not as a string.

File: plugins/inventory/iocage.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            description:
+              - Get jails' properties.
+                Creates dictionary C(iocage_properties) for each added host.
+            type: boolean
+            default: false
+        env:
+            description: O(user)'s environment on O(host).
+            type: list
+            elements: str
+            default: []
```
Comment (russoz) [inline plugins/inventory/iocage.py:55]: I would argue that it makes more sense to have a `dict` here rather than `list`

File: plugins/inventory/iocage.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+        cmd_list = cmd.copy()
+        cmd_list.append('list --header --long')
+        cmd_list = ' '.join(cmd_list)
+        try:
+            p = Popen(cmd_list, shell=True, text=True, stdout=PIPE, stderr=PIPE)
+            stdout, stderr = p.communicate()
+            if p.returncode != 0:
+                raise AnsibleParserError('Failed to run cmd=%s, rc=%s, stderr=%s' %
+                                         (cmd_list, p.returncode, to_native(stderr)))
```
Comment (russoz) [inline plugins/inventory/iocage.py:190]: Arguably that it not a parser error. That looks more like `AnsibleError` than `...ParserError`

**Review by russoz — COMMENTED** [review]
Hi @vbotka thanks for your contribution

Couple of issues/suggestions.

File: plugins/inventory/iocage.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+# user is not required if iocage is running on localhost
+plugin: community.general.iocage
+host: localhost
+
+# run cryptography without legacy algorithms
+plugin: community.general.iocage
+host: 10.1.0.73
+user: admin
+env:
+  - CRYPTOGRAPHY_OPENSSL_NO_LEGACY=1
```
[the remark this answers is not in the document]
Comment (vbotka) [reply plugins/inventory/iocage.py:82]: This is an example of how to use the *env* parameter. To 'run cryptography without legacy algorithms` in this particular case. An example can't declare any parameter default. The default of *env* is '[]' (line 55.)

File: plugins/inventory/iocage.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            raise AnsibleParserError('Failed to parse %s: %s. iocage_data: %s' %
+                                     (to_native(path), to_native(e), to_native(iocage_data)))
+
+        if get_properties:
+            for hostname, host_vars in results['_meta']['hostvars'].items():
+                cmd_get_properties = cmd.copy()
+                cmd_get_properties.append(f"get --all {hostname}")
+                cmd_get_properties = ' '.join(cmd_get_properties)
+                try:
+                    p = Popen(cmd_get_properties, shell=True, text=True, stdout=PIPE, stderr=PIPE)
```
[the remark this answers is not in the document]
Comment (vbotka) [reply plugins/inventory/iocage.py]: OK.Done.

File: plugins/inventory/iocage.py (hunk 1 above)
Comment (vbotka) [reply plugins/inventory/iocage.py]: OK.Done.

File: plugins/inventory/iocage.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+              - The name of this plugin, it should always be set to
+                V(community.general.iocage) for this plugin to recognize
+                it as its own.
+            required: true
+            choices: ['community.general.iocage']
+            type: str
+        host:
+            description: The ip/name of the C(iocage) host.
+            required: true
+            type: str
```
[the remark this answers is not in the document]
Comment (vbotka) [reply plugins/inventory/iocage.py]: Good point! Done.

File: plugins/inventory/iocage.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        env = self.get_option('env')
+        get_properties = self.get_option('get_properties')
+
+        cmd = []
+        if host != 'localhost':
+            user = self.get_option('user')
+            cmd.append("ssh")
+            cmd.append(f"{user}@{host}")
+        if env:
+            cmd.extend(env)
```
Comment (russoz) [inline plugins/inventory/iocage.py:179]: Not sure this is going to work as intended now. `Popen` constructor accepts a parameter `env` which is a `dict` (I swear I wasn't thinking of this when I suggested you change the data type for that param, but then again, it is inherently a mapping of keys to values).

Comment (vbotka) [reply plugins/inventory/iocage.py:179]: It's working. I've already tested it. The `Popen` parameter `env` would create the environment for `ssh`. If the plugin parameter `env` is set it must set the environment for `iocage`.

File: plugins/inventory/iocage.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+            type: boolean
+            default: false
+        env:
+            description: O(user)'s environment on O(host).
+            type: list
+            elements: str
+            default: []
+    notes:
+      - You might want to test the command C(ssh user@host iocage list -l)
+        on the controller before using this inventory plugin.
```
Comment (russoz) [inline plugins/inventory/iocage.py:58]: I believe it is important to let the users know that this plugin does not allow specifying passwords nor encryption keys to perform the ssh operation on the remote node, and they should make use of `~/.ssh/config` to control those and other settings.
Also, I am thinking that some users might assume, mistakenly, that the ssh aspect of this will be handled by Ansible itself (as it uses ssh by default to reach remote targets).

TBH, I have mixed feelings about this usage of ssh, but I do not have a better way to do this. I cannot help thinking that if I were to use this, I might prefer making it in two steps: one ansible playbook to fetch the file from the remote, another one actually using it. As I wrote this, I recalled there was a case in `ignore-X.X.txt` files about a plugin that used a module: `iptables_state` action plugin _uses_ the `iptables_state` module. I have never fully understood this, nor have I spent time trying to, but it just _might_ provide you with a mechanism to let Ansible handle the ssh part for you. 

@felixfontein any insights on this?

File: plugins/inventory/iocage.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (russoz) [reply plugins/inventory/iocage.py:81]: Yup, I got that, my recommendation was for the module logic, but it might be a bit overkill for this plugin.

File: plugins/inventory/iocage.py (hunk 3 above)
Comment (vbotka) [reply plugins/inventory/iocage.py:191]: Why do `nmap.py` and `virtualbox.py` use `AnsibleParserError` to handle `Popen` errors?

File: plugins/inventory/iocage.py (hunk 8 above)
Comment (vbotka) [reply plugins/inventory/iocage.py:58]: > ... I might prefer making it in two steps: one ansible playbook to fetch the file from the remote, another one actually using it.

If you prefer a playbook use the role [vbotka.freebsd_iocage](https://galaxy.ansible.com/ui/standalone/roles/vbotka/freebsd_iocage/).

Comment (vbotka) [reply plugins/inventory/iocage.py:58]: >  ... some users might assume, mistakenly, that the ssh aspect of this will be handled by Ansible 

I agree. Where would you like to put it? There are only the notes below atm.

```yaml
    notes:                                                                                                           
      - You might want to test the command C(ssh user@host iocage list -l)                                           
        on the controller before using this inventory plugin.                                                        
      - If you run this inventory plugin on V(localhost) C(ssh) is not used.                                         
        In this case, test the command C(iocage list -l).
        ...
```

Comment (vbotka) [reply plugins/inventory/iocage.py:58]: > ... it just _might_ provide you with a mechanism to let Ansible handle the ssh part for you.

Thank you for the effort. I'd like to keep it as simple as possible. By default, you don't need root to run *iocage* and get the data. There might be use cases where mandatory Ansible access would be insecure. Just like with any other VM manager. You get access to the VMs not to the manager.

File: plugins/inventory/iocage.py (hunk 7 above)
Comment (russoz) [reply plugins/inventory/iocage.py:179]: Ok, makes sense.

File: plugins/inventory/iocage.py (hunk 3 above)
Comment (russoz) [reply plugins/inventory/iocage.py:191]: Most likely because I wasn't there ;-) Or I wasn't paying attention :-) Anyways, regardless of that, would you agree that this is not a parsing issue?

File: plugins/inventory/iocage.py (hunk 2 above)
Comment (russoz) [reply plugins/inventory/iocage.py:55]: The argument would be that environment variables are, inherently, a mapping construct, they map var names to their respective values. That's the reason why `Popen`'s env argument expects a dict, not a list, same for `os.environ`. And I would assume that most, if not all, modules and plugins treat sets of environment vars as a dict rather than a list, so another potential argument would be consistency.

That being said, I'm not planning on pressing this point. If you still want to keep it as a list, feel free to.

File: plugins/inventory/iocage.py (hunk 3 above)
Comment (vbotka) [reply plugins/inventory/iocage.py]: You're right. It's not a parsing issue. I fixed the calls.

File: plugins/inventory/iocage.py (hunk 7 above)
Comment (vbotka) [reply plugins/inventory/iocage.py:179]: Then, this resolves also "Comment on lines +53 to +55".

File: plugins/inventory/iocage.py (hunk 2 above)
Comment (vbotka) [reply plugins/inventory/iocage.py:55]: Yes. I want to keep it as a list. I oppose your arguments:

1) The parameter `env` in this plugin can't accept a variable.
2) We already clarified that `Popen` argument `env` can't be used here.
3) `shell> grep -r 'Popen' plugins/ | grep 'env'` shows nothing.
4) The consistency is a valid point. Limited by the plugin context.

In summary, I argue that the consistency here is not worth the dictionary complexity burden.

File: plugins/inventory/iocage.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        plugin:
+            description:
+              - The name of this plugin, it should always be set to
+                V(community.general.iocage) for this plugin to recognize
+                it as its own.
+            required: true
+            choices: ['community.general.iocage']
+            type: str
+        host:
+            description: The ip/name of the C(iocage) host.
```
Comment (felixfontein) [inline plugins/inventory/iocage.py:35]:
```suggestion
            description: The IP/hostname of the C(iocage) host.
```

File: plugins/inventory/iocage.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+            type: str
+            default: localhost
+        user:
+            description:
+              - C(iocage) user.
+                It is expected that the O(user) is able to connect to the
+                O(host) and execute the command C(iocage list).
```
Comment (felixfontein) [inline plugins/inventory/iocage.py:42]:
```suggestion
                O(host) with SSH and execute the command C(iocage list).
```

File: plugins/inventory/iocage.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    short_description: iocage inventory source
+    version_added: 10.2.0
+    author:
+        - Vladimir Botka (@vbotka)
+    requirements:
+        - iocage >= 1.8
+    description:
+        - Get inventory hosts from the iocage jail manager.
+        - Uses a configuration file as an inventory source, it must end
+          in C(.iocage.yml) or C(.iocage.yaml).
```
Comment (felixfontein) [inline plugins/inventory/iocage.py:21]: I think description should mention that remote hosts are contacted via SSH. From the description I would assume that the plugin simply runs iocage locally.

File: plugins/inventory/iocage.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+    extends_documentation_fragment:
+        - constructed
+        - inventory_cache
```
Comment (felixfontein) [inline plugins/inventory/iocage.py:24]:
```suggestion
        - ansible.builtin.constructed
        - ansible.builtin.inventory_cache
```

File: plugins/inventory/iocage.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+            type: list
+            elements: str
+            default: []
+    notes:
+      - You might want to test the command C(ssh user@host iocage list -l)
+        on the controller before using this inventory plugin.
+      - If you run this inventory plugin on V(localhost) C(ssh) is not used.
+        In this case, test the command C(iocage list -l).
+      - This inventory plugin creates variables C(iocage_*) for each added host.
+      - The values of these variables are collected from the output of the command C(iocage list -l)
```
Comment (felixfontein) [inline plugins/inventory/iocage.py:62]:
```suggestion
      - The values of these variables are collected from the output of the command C(iocage list -l).
```

File: plugins/inventory/iocage.py (hunk 8 above)
Comment (felixfontein) [inline plugins/inventory/iocage.py:58]:
```suggestion
        on the controller before using this inventory plugin with O(user) specified and with O(host) other than V(localhost).
```

File: plugins/inventory/iocage.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+            p = Popen(cmd_list, stdout=PIPE, stderr=PIPE)
+            stdout, stderr = p.communicate()
+            if p.returncode != 0:
+                raise AnsibleError('Failed to run cmd=%s, rc=%s, stderr=%s' %
+                                   (cmd_list, p.returncode, to_native(stderr)))
+
+            try:
+                t_stdout = to_text(stdout, errors='surrogate_or_strict')
+            except UnicodeError as e:
+                raise AnsibleError('Invalid (non unicode) input returned: %s' % to_native(e))
```
Comment (felixfontein) [inline plugins/inventory/iocage.py:196]:
```suggestion
                raise AnsibleError('Invalid (non unicode) input returned: %s' % to_native(e)) from e
```

File: plugins/inventory/iocage.py [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+                results['_meta']['hostvars'][iocage_name]['iocage_type'] = jail[4]
+                results['_meta']['hostvars'][iocage_name]['iocage_release'] = jail[5]
+                results['_meta']['hostvars'][iocage_name]['iocage_ip4'] = _parse_ip4(jail[6])
+                results['_meta']['hostvars'][iocage_name]['iocage_ip6'] = jail[7]
+                results['_meta']['hostvars'][iocage_name]['iocage_template'] = jail[8]
+                results['_meta']['hostvars'][iocage_name]['iocage_basejail'] = jail[9]
+
+        except Exception as e:
+            raise AnsibleParserError('Failed to parse %s: %s. iocage_data: %s' %
+                                     (to_native(path), to_native(e), to_native(iocage_data)))
```
Comment (felixfontein) [inline plugins/inventory/iocage.py:216]:
```suggestion
                                     (to_native(path), to_native(e), to_native(iocage_data))) from e
```
Also `iocage_data` is not necessarily defined at this point, for example if the above `AnsibleError`s have been raised.

File: plugins/inventory/iocage.py (hunk 8 above)
Comment (felixfontein) [reply plugins/inventory/iocage.py:58]: > And as I wrote it, I recalled there was a case in ignore-X.X.txt files about a plugin that used a module: iptables_state action plugin uses the iptables_state module. I have never fully understood this, nor have I spent time trying to, but it just might provide you with a mechanism to let Ansible handle the ssh part for you.

Action plugins are special because they are the controller version of modules, and users cannot distinguish between modules and action plugins (and usually action plugins come with modules, if only for documentation). So this is not related to inventory plugins at all :)

Comment (felixfontein) [reply plugins/inventory/iocage.py:58]: > TBH, I have mixed feelings about this use of ssh, but I do not have a better way to do this. I cannot help thinking that if I were to use this, I might prefer making it in two steps: one ansible playbook to fetch the file from the remote, another one actually using it.

I agree. It would be nice to be able to stack plugins (connection plugins come to mind here, but also become plugins, and inventory plguins could use both) - but that isn't possible at the moment. I think I have seen at least one attempt of a plugin trying to load another plugin for this reason, but that's something I wouldn't recommend - at least for now. It might be worth trying this out in another collection (that's less stable than c.g) to see whether it's a direction worth pursuing...

File: tests/unit/plugins/inventory/test_iocage.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+@pytest.fixture(scope="module")
+def inventory():
+    return InventoryModule()
+
+
+def test_verify_file(tmp_path, inventory):
+    file = tmp_path / "foobar.iocage.yml"
+    file.touch()
+    assert inventory.verify_file(str(file)) is True
```
Comment (felixfontein) [inline tests/unit/plugins/inventory/test_iocage.py:23]: It would be nice to have more extensive tests, than this absolutely bare minimum :)

Comment (vbotka) [reply tests/unit/plugins/inventory/test_iocage.py:23]: I'm working on it.

File: plugins/inventory/iocage.py (hunk 15 above)
Comment (vbotka) [reply plugins/inventory/iocage.py]: I removed iocage_data from the AnsibleParserError. Thank you!

File: plugins/inventory/iocage.py (hunk 11 above)
Comment (vbotka) [reply plugins/inventory/iocage.py:24]: OK. I added an item to the description.

File: tests/unit/plugins/inventory/test_iocage.py (hunk 16 above)
Comment (vbotka) [reply tests/unit/plugins/inventory/test_iocage.py]: I completed the tests.

File: plugins/inventory/iocage.py (hunk 8 above)
Comment (russoz) [reply plugins/inventory/iocage.py]: > > ... some users might assume, mistakenly, that the ssh aspect of this will be handled by Ansible
> 
> The parameter _user_ says:
> 
> ```yaml
>         user:
>             description:
>               - C(iocage) user.
>                 It is expected that the O(user) is able to connect to the
>                 O(host) and execute the command C(iocage list).
>                 This option is not required if O(host) is V(localhost).
> ```
> 
> Also,
> 
> ```yaml
>     notes:                                                                                                           
>       - You might want to test the command C(ssh user@host iocage list -l)                                           
>         on the controller before using this inventory plugin.                                                        
>       - If you run this inventory plugin on V(localhost) C(ssh) is not used.                                         
>         In this case, test the command C(iocage list -l).
>         ...
> ```

Not going to press the point, however I still think it is worth to mention that this ssh must work without interaction (e.g. keys without passphrase and/or other .ssh/config gimmicks). Again, it's your call.

File: plugins/inventory/iocage.py [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+    version_added: 10.2.0
+    author:
+        - Vladimir Botka (@vbotka)
+    requirements:
+        - iocage >= 1.8
+    description:
+        - Get inventory hosts from the iocage jail manager running on O(host).
+        - By default, O(host) is V(localhost). If O(host) is not V(localhost) it
+          is expected that the O(user) is able to connect to the O(host) with
+          SSH and execute the command C(iocage list).
```
Comment (russoz) [inline plugins/inventory/iocage.py:22]: Technically, it is not the `O(user)` (remote user on the target iocage host) that must be able to connect. This would be more accurate as:
	"... expected that the user running Ansible on the controller is able to connect with SSH to the O(host) with O(user) without specifying password or key, and O(user) is able to execute the command C(iocage list)."

Or something along those lines - feel free to wordsmith it to your satisfaction.

File: plugins/inventory/iocage.py (hunk 8 above)
Comment (vbotka) [reply plugins/inventory/iocage.py]: Thank you. I added `non-interactively` to the description item

```
        - By default, O(host) is V(localhost). If O(host) is not V(localhost) it
          is expected that the O(user) is able to connect to the O(host) with
          SSH non-interactively and execute the command C(iocage list).
```

File: plugins/inventory/iocage.py (hunk 17 above)
Comment (vbotka) [reply plugins/inventory/iocage.py]: Thank you. I updated the description item

```
        - By default, O(host) is V(localhost). If O(host) is not V(localhost) it
          is expected that the user running Ansible on the controller can
          connect to the O(host) account O(user) with SSH non-interactively and
          execute the command C(iocage list).
```

**Review by russoz — APPROVED** [review]
LGTM

File: plugins/inventory/iocage.py (hunk 17 above)
Comment (russoz) [reply plugins/inventory/iocage.py]: Thanks!

File: plugins/inventory/iocage.py [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+        env = self.get_option('env')
+        get_properties = self.get_option('get_properties')
+
+        cmd = []
+        if host != 'localhost':
+            user = self.get_option('user')
+            cmd.append("ssh")
+            cmd.append(f"{user}@{host}")
+        if env:
+            cmd.extend(env)
```
Comment (felixfontein) [inline plugins/inventory/iocage.py:185]: This does not work if `host=localhost`. The first element of `cmd` must be the executable that is run. You can simply check that by running
```
python -c 'from subprocess import Popen, PIPE ; print(Popen(["A=B", "/bin/sh", "-c", "export"], stdout=PIPE, stderr=PIPE).communicate())'
```
on the command line. This results in
```
FileNotFoundError: [Errno 2] No such file or directory: 'A=B'
```
You must use the `env` parameter to Popen (https://docs.python.org/3/library/subprocess.html#subprocess.Popen) for `host=localhost`. (Which needs to be a dictionary, so having `env` as a dictionary would definitely be better.)

Comment (vbotka) [reply plugins/inventory/iocage.py]: Right. I fixed `env`.

File: plugins/inventory/iocage.py [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+
+    def get_inventory(self, path):
+        host = self.get_option('host')
+        env = self.get_option('env')
+        get_properties = self.get_option('get_properties')
+
+        cmd = []
+        my_env = os.environ.copy()
+        if host == 'localhost':
+            my_env = my_env | env
```
[the remark this answers is not in the document]
Comment (vbotka) [reply plugins/inventory/iocage.py]: OK. Done. (JFYI, I tested quoted values and forgot to update examples.) The conversions are more robust. Thank you!

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@vbotka thanks for your contribution!
@russoz thanks for reviewing!

**Comment by vbotka** [comment]
@felixfontein, @russoz thank you! May I close #9261?

**Comment by felixfontein** [comment]
Sure, feel free to close it.
